cpp/src/parquet/column_writer.cc

```diff
@@ -1014,11 +1014,58 @@ template <typename Action>
 inline void DoInBatches(int64_t total, int64_t batch_size, Action&& action) {
   int64_t num_batches = static_cast<int>(total / batch_size);
   for (int round = 0; round < num_batches; round++) {
-    action(round * batch_size, batch_size);
+    action(round * batch_size, batch_size, /*check_page=*/true);
   }
   // Write the remaining values
   if (total % batch_size > 0) {
-    action(num_batches * batch_size, total % batch_size);
+    action(num_batches * batch_size, total % batch_size, /*check_page=*/true);
+  }
+}
+
+template <typename Action>
+inline void DoInBatches(const int16_t* def_levels, const int16_t* rep_levels,
+                        int64_t num_levels, int64_t batch_size, Action&& action,
```

Comment:
Yes, `num_levels` is the length of `def_levels` and `rep_levels`. It is used by various functions in this file so it would be better to be consistent.
```cpp
  Status WriteArrowDictionary(const int16_t* def_levels, const int16_t* rep_levels,
                              int64_t num_levels, const ::arrow::Array& array,
                              ArrowWriteContext* context, bool maybe_parent_nulls);

  Status WriteArrowDense(const int16_t* def_levels, const int16_t* rep_levels,
                         int64_t num_levels, const ::arrow::Array& array,
                         ArrowWriteContext* context, bool maybe_parent_nulls);
```

IMHO, "depth" is the value at each slot of `def_levels` and `rep_levels`. So the name `num_levels` makes sense to me. WDYT?

Reply:
+1 for consistency

Reply:
Okay that's fair. This is definitely a pre-existing issue, but I wish we used something like `std::span` instead of separate pointers and lengths everywhere. Would avoid these kind of naming issues.

Reply:
These pointers are optional and already used as public apis. Changing them into span is a braking change. That can be a separate patch and needs more discussion. 

Do you have any other concerns? Or it can be merged first to unblock the page index impl in another PR of mine? @wjones127

Reply:
This is fine to merge as is.